README.md
![Plock Logo](./assets/cover.png)

React Plock is a tree-shakeable **ultra small** npm package (**less than 1kB gzipped**) that allows you to create amazing masonry layouts with an amazing developer experience. With React Plock, you can easily create responsive and customizable layouts that adapt to different screen sizes and devices.

### Features

- **Masonry Layout**: Create beautiful masonry layouts with ease.
- **Responsive**: Automatically adapts to different screen sizes and devices.
- **Customizable**: Customize the layout to match your needs.
- **TypeScript Ready**: Get the strength of type-safe languages.
- **Amazing DX**: Easy to use and well-documented.

### Installation

```bash
npm install react-plock
```

### Usage

Using Plock with the new v3 APIs it's a piece of cake. Here's an example of how can you create an [Unsplash-Like](https://unsplash.com/) masonry grid. You can even see a demo of this example by clicking [here](https://react-plock.netlify.app/).

```tsx
import { Masonry } from 'react-plock';

const ImagesMasonry = () => {
  const items = [...imageUrls];

  return (
    <Masonry
      items={items}
      config={{
        columns: [1, 2, 3],
        gap: [24, 12, 6],
        media: [640, 768, 1024],
      }}
      render={(item, idx) => (
        <img key={idx} src={item} style={{ width: "100%", height: "auto" }} />
      )}
    />
  );
};
```

### API Reference

Here's the TypeScript definition for the Masonry Component, below you can find a more detailed explanation.

```ts
export type MasonryProps<T> = React.ComponentPropsWithoutRef<"div"> & {
  items: T[];
  render: (item: T, idx: number) => React.ReactNode;
  config: {
    columns: number | number[];
    gap: number | number[];
    media?: number[];
  };
};
```

#### Items

This prop accepts a generic array of elements, each one will be passed to the **render** property.

#### Render

The masonry render prop. Here's where you define the styles of every tile of the grid, the function takes the current looping item and the relative index.

#### Config

A configuration object that is used to define the number of columns, media queries and gaps between items.

#### Other Props

As you can see, by using `React.ComponentPropsWithoutRef<"div">` you can simply pass every available property to the div, some examples are **id** and **className**. The only one property that will be overwritten will be the `style` because is used internally for the masonry generation.

### Important Note

Please, note that in case you are passing an array to the columns attribute of the config property, the number of elements **MUST** be equal to the number of media AND gap breakpoints provided!

```tsx
// Correct: This will be responsive with 3 breakpoints.
<Masonry
  {...otherProps}
  config={{
    columns: [1, 2, 3],
    gap: [12, 16, 20],
    media: [640, 768, 1024],
  }}
/>

// Correct: This will be responsive with 2 breakpoints.
<Masonry
  {...otherProps}
  config={{
    columns: [1, 2],
    gap: [2, 4],
    media: [640, 1024],
  }}
/>

// Correct: This will be fixed with 4 columns in every screen size.
<Masonry
  {...otherProps}
  config={{
    columns: 4,
    gap: 8
  }}
/>

// NOT Correct: This will cause trouble in rendering.
<Masonry
  {...otherProps}
  config={{
    columns: [4],
    media: [640, 768],
  }}
/>
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 3.0.2 (2023-03-27)

**Note:** Version bump only for package root





## 3.0.2-beta.1 (2023-03-27)

**Note:** Version bump only for package root





## 3.0.2-beta.0 (2023-03-27)

**Note:** Version bump only for package root





## 3.0.1 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.7 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.6 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.5 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.4 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.3 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.2 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.1 (2023-03-24)

**Note:** Version bump only for package root





## 3.0.1-beta.0 (2023-03-24)

**Note:** Version bump only for package root


src/components/Plock.js
import * as React from "react";

/**
 * Configuration for Plock.
 * This is a map of breakpoints to the number of columns to use for that breakpoint.
 *
 * const breakpoints = [
 *   { size: 640, columns: 1 },
 *   { size: 768, columns: 2 },
 *   { size: 1024, columns: 3 },
 *   { size: 1280, columns: 4 },
 * ];
 */

// Check if window is defined (so if in the browser or in node.js).
const isBrowser = typeof window !== "undefined";

export function useWindowWidth({ debounceMs }) {
  const [width, setWidth] = React.useState(isBrowser ? window.innerWidth : 0);
  const handleResize = useDebounce(
    () => setWidth(isBrowser ? window.innerWidth : 0),
    debounceMs
  );

  React.useEffect(() => {
    if (isBrowser) {
      window.addEventListener("resize", handleResize);
      return () => window.removeEventListener("resize", handleResize);
    }
  }, [handleResize]);

  return width;
}

export function useDebounce(fn, ms) {
  let timeout = null;

  return (...args) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => fn(args), ms);
  };
}

export const Plock = React.forwardRef(
  (
    {
      as: Comp = "div",
      children,
      className,
      style,
      gap = 10,
      debounce = 200,

      /**
       * TODO:
       * This will be renamed to breakpoints in a future major release!
       */
      nColumns: breakpoints = 3,
    },
    forwardedRef
  ) => {
    const width = useWindowWidth({ debounceMs: debounce });
    const [columns, setColumns] = React.useState([]);

    React.useLayoutEffect(() => {
      const first = (breakpoints) => {
        return breakpoints?.[0];
      };

      const last = (breakpoints) => {
        return breakpoints?.[breakpoints.length - 1];
      };

      const sorted = (breakpoints) => {
        return breakpoints.sort((a, b) => a.size - b.size);
      };

      const contained = (breakpoints, width) => {
        return breakpoints.filter((el) => el.size <= width);
      };

      const isNumber = (element) => typeof element === "number";

      const breakpoint = isNumber(breakpoints)
        ? { columns: breakpoints }
        : last(sorted(contained(breakpoints, width))) ??
          first(sorted(breakpoints));

      const columnsForBreakpoint = Array.from(
        { length: breakpoint.columns },
        (e) => []
      );

      React.Children.forEach(children, (child, index) => {
        const key = `item-${index}`;
        const cloned = React.cloneElement(child, {
          ...child.props,
          key: key,
        });

        columnsForBreakpoint[index % columnsForBreakpoint.length].push(cloned);
      });

      setColumns(columnsForBreakpoint);
    }, [children, breakpoints, width]);

    const defaultStyles = {
      mainGrid: {
        display: "grid",
        gridTemplateColumns: `repeat(${columns.length}, 1fr)`,
        columnGap: gap,
        alignItems: "start",
      },
      columnGrid: {
        display: "grid",
        gridTemplateColumns: "100%",
        rowGap: gap,
      },
    };

    return (
      <Comp
        ref={forwardedRef}
        data-testid="plock-container"
        className={className}
        style={{ style, ...defaultStyles.mainGrid }}
      >
        {columns.map((column, index) => {
          return (
            <div
              data-testid="plock-column"
              style={defaultStyles.columnGrid}
              key={index}
            >
              {column}
            </div>
          );
        })}
      </Comp>
    );
  }
);
